vignettes/query.Rmd

```diff
@@ -0,0 +1,212 @@
+`usedby` can be combined with groupings and scope:
+
+```r
+outpack_search(quote(usedby(latest(name == "C")) && parameter:year == 2022),
+               name = "A")
+```
+
+The depth that `usedby` will recurse can be controlled by setting the `depth` e.g.
+
+```r
+outpack_search(quote(usedby(latest(name == "C"), depth = 1)), name = "A")
+```
+
+will search for just immediate parents of `C`. `depth` can be any positive integer, by default `depth` will recurse until it finds all parents.
+
+`usedby` can be simplified by using subqueries. Subqueries are denoted by curly braces `{}` and can either be named and passed in `subquery` arg or can be anonymous. The query below is equivalent to the above but uses a subquery for `C`.
+
+```r
+outpack_search(quote(usedby({C}) && parameter:year == 2022),
+               name = "A",
+               subquery = list(C = quote(latest(name == "C"))))
+```
+
+There are two important things to note about `usedby`:
+* The query inside `usedby` will search the entire index, ignoring any `scope` or `name` parameters. This is because we want to find all packets which are used by `latest` `C`. If the subquery `C` was scoped this would return no results.
+* The query inside `usedby` must return a single result. To ensure this it must either be a literal `id`, a call to `latest` or a call to `single`
+
+As well as searching up the dependency tree using `usedby` we can search down with the `uses` function. In the same setup above with reports `A`, `B` and `C` if we want to know the `id` of `C` which uses `A` we can find this by using
+
+```r
+outpack_search(quote(uses(latest(name == "A"))),
+               name = "C")
+```
+
+`uses` and `useby` can be combined to search more complex arrangements of dependencies. If we have something like
+
+```{r, echo = FALSE, fig.align = 'center'}
+knitr::include_graphics("images/dependency_multiple.png")
+```
+
+If we want to search for the version of `E` which depends on the version of `A` which was used in the latest `C` we can do this via
+
+```
+outpack_search(
+  quote(latest(uses(single(usedby(latest(name == "C")) && name == "A")))),
+  name = "E")
+```
+
+This searches up the tree from `C` to `A` and then down the tree to find the version of `E`. Note that is is important we added the `name == "A"` condition here, if that was missing `usedby(latest(name == "C"))` would also return `B` and `single` would throw an error because we have multiple packets.
+
+We can also search up the tree and then down to find `A` from `D` e.g.
+
+```
+outpack_query(
+  quote(usedby(single(uses(name == "D")))),
+  name = "A",
+  root = root)
+```
+
+note as `E` is the only packet which uses `D` we do not need to add a `name == "E"` clause.
+
+We can combine `usedby` and `uses` in more complex searches, such as to find `D` from `C` we could run
+
+```
+outpack_query(
+  quote(usedby(single(uses(single(usedby(latest(name == "C")) && name == "A"))) && name == "E"))),
+  name = "D",
+  root = root)
+```
+
```

Comment:
I almost feel there should be some practice exercises.

Also - are circular dependencies worth mentioning - presumably this isn't allowed?